homeassistant/components/fivem/config_flow.py

```diff
@@ -0,0 +1,76 @@
+"""Config flow for FiveM integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from fivem import FiveM, FiveMServerOfflineError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_PORT
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+import homeassistant.helpers.config_validation as cv
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_PORT = 30120
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME): cv.string,
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    }
+)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> None:
+    """Validate the user input allows us to connect."""
+
+    fivem = FiveM(data[CONF_HOST], data[CONF_PORT])
+    info = await fivem.get_info_raw()
+
+    gamename = info.get("vars")["gamename"]
+    if gamename is None or gamename != "gta5":
+        raise InvalidGamenameError
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for FiveM."""
+
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
+            )
+
+        errors = {}
+
+        try:
+            await validate_input(self.hass, user_input)
+        except FiveMServerOfflineError:
+            errors["base"] = "cannot_connect"
+        except InvalidGamenameError:
+            errors["base"] = "invalid_gamename"
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            errors["base"] = "unknown"
+        else:
+            return self.async_create_entry(title=user_input[CONF_NAME], data=user_input)
```

Comment:
We should probably guard against creating an entry with the same host and port combo.

We have this helper:
https://github.com/home-assistant/core/blob/d12a3927678296bd87852b1c660ca511813c2247/homeassistant/config_entries.py#L1207-L1210